### PR TITLE
Apply upstream patch 3842: Linker: Object unloading

### PR DESCRIPTION
**Pull request by rayshih**
Summary:

This commit applies

https://gitlab.haskell.org/ghc/ghc/-/merge_requests/3842

except this still uses HashTable instead of StrHashTable

This commit also includes changes done in this commit:

https://gitlab.haskell.org/ghc/ghc/-/commit/7be3df495f49d5faaf137f21e09edf00a980c5ef

Worth to notice that following commits need to be reapplied.

- https://github.com/facebook/fbghc/commit/74d8cadeef8a509f1ee612795f5af6142edb006c
- https://github.com/facebook/fbghc/commit/6217823b08ac0b5b8e3bfef4265a8cd46456d253

For easier review, I'll put those in another PR.

Test Plan:

./validate

and all passed

Reviewers: josefs

### CODE REVIEW

File: testsuite/tests/rts/linker_unload_native.c [hunk 1]
```diff
@@ -41,25 +40,14 @@ int main (int argc, char *argv[])
...
-        // loading the same library again should fail
-        void* handle1 = loadNativeObj(OBJPATH, &errmsg);
-        if (!handle1) {
-            // don't spam the output file:
-            // errorBelch("loadNativeObj(%s) failed: %s", OBJPATH, errmsg);
-            free(errmsg);
-        } else {
-            errorBelch("loadNativeObj(%s) should have failed", OBJPATH);
-            exit(1);
-        }
```
Comment (rayshih) [inline testsuite/tests/rts/linker_unload_native.c]: this will be added back in next PR

Comment (rayshih) [reply testsuite/tests/rts/linker_unload_native.c]: done, recommend to review by the commit itself only

File: rts/Linker.c [hunk 2]
```diff
@@ -1901,6 +1901,16 @@ static void * loadNativeObj_ELF (pathchar *path, char **errmsg)
 
    IF_DEBUG(linker, debugBelch("loadNativeObj_ELF %" PATH_FMT "\n", path));
 
+   // Loading the same object multiple times will lead to chaos
+   // because we will have two NativeCodes but one underlying handle,
+   // so let's fail if this happens.
+   for (nc = loaded_objects; nc; nc = nc->next) {
```
Comment (pepeiborra) [inline rts/Linker.c:1907]: where does `loaded_objects` come from? I cannot find the definition site

**Review by pepeiborra — APPROVED** [review]
I have reviewed the second commit and it looks good to me

**Pull request merged**